Fixed setting pie plot options with multiple data labels renders incorrectly

Eskils opened the pull request:
Fixed #21928, setting plot options for pie chart with multiple data labels did not work.

- Issue was resolved by merging data label options in pie plot options with the defaults. 
   + This was not already happening because the `merge` function does not merge arrays.
- Updated pie demos with multiple data labels to use pie plot options instead of series plot options. 
   + This seems more correct since the demos will not compile in TypeScript.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208435267040163

goransle commented on ts/Core/Chart/Chart.ts:
```diff
@@ -434,6 +436,10 @@ class Chart {
             const options = merge(defaultOptions, userOptions), // Do the merge
                 optionsChart = options.chart;
 
+            if (options.plotOptions.pie) {
+                this.mergePieDataLabelPlotOptions(options.plotOptions.pie);
+            }
```
This should likely happen in `PieSeries.ts`. This seems to work:

```ts
addEvent(PieSeries, 'afterInit', (e): void => {
    const chart = e.target.chart;
    if (chart.options.plotOptions.pie) {
        PieSeries.mergePieDataLabelPlotOptions(chart.options.plotOptions.pie);
    }
});
```

Possibly the `update` or `setOptions` events can be used for processing when changing options.

goransle commented on samples/unit-tests/series-pie/datalabel/demo.js:
```diff
@@ -652,3 +652,92 @@ QUnit.test('Connector color of individual point (#8864).', function (assert) {
+            ]
+        });
+
+        assert.ok(
```
I would slightly prefer `assert.strictEqual` for test failure reporting

Eskils commented:
Thank you for your feedback! The `setOptions` event was what I needed.

goransle reviewed (approved):
Nice, looks good!

hubertkozik commented:
I think this problem is deeper than only for `PieSeries`. I tried to use the same config for the line series: https://jsfiddle.net/BlackLabel/a8grfdvc/ and sadly the error came up. That would mean, that your fix is not enough and we should fix it somewhere deeper in the code 😕

hubertkozik reviewed (changes requested):
I think this problem is deeper than only for `PieSeries`. I tried to use the same config for the line series: https://jsfiddle.net/BlackLabel/a8grfdvc/ and sadly the error came up. That would mean, that your fix is not enough and we should fix it somewhere deeper in the code 😕

Eskils commented:
> I think this problem is deeper than only for `PieSeries`. I tried to use the same config for the line series: https://jsfiddle.net/BlackLabel/a8grfdvc/ and sadly the error came up. That would mean, that your fix is not enough and we should fix it somewhere deeper in the code 😕

Thank you for testing more broadly—good find! I added a unit-test for line-series and moved my patch to Series.ts.

hubertkozik commented on samples/unit-tests/series-line/datalabel/demo.js:
Do we need this new file and this QUnit test instance? Those instances are quite heavy, it is better and we are trying to add some code to the existing test because updates are lighter than creating new tests and new chart instances. Can you put it somewhere in existing test?

Eskils replied:
Good point. I integrated the _series-line/datalabel_ test in the existing _series-line/linewidth_ test. This greatly reduced the test run duration.

The new tests in _series-pie/datalabel_ are already quite swift, and integrating them into existing tests had little to no effect so they are left unchanged (See attached image).

<img width="836" alt="Screenshot 2024-11-11 at 17 37 10" src="https://github.com/user-attachments/assets/5918509d-9785-403f-a1e6-8e0e8d86dd57">

hubertkozik replied:
Nice! Thank you.

hubertkozik commented on ts/Core/Series/Series.ts:
```diff
@@ -816,6 +816,19 @@ class Series {
+        // When plot options have multiple data labels, the default data label
+        // plot options are not merged (#21928).
+        const dataLabelOptions = typeOptions.dataLabels;
+        if (dataLabelOptions instanceof Array) {
```
I'm thinking about using `splat` function to simplify logic a little bit. Did you consider using `splat` and operating always on the array?

hubertkozik replied:
Right, got it!

hubertkozik reviewed (commented):
Sorry for the long wait for review! The notification got lost somehow :(

TorsteinHonsi commented on ts/Core/Series/Series.ts:
```diff
@@ -816,6 +816,19 @@ class Series {
             ),
             userPlotOptionsType = userPlotOptions[this.type] || {};
 
+        // When plot options have multiple data labels, the default data label
+        // plot options are not merged (#21928).
```
This comment describes a situation that is no longer true. Instead, describe what the code actually does, and add a reference to the issue.
```suggestion
        // Merge in multiple data label options from the plot option (#21928).
```

TorsteinHonsi reviewed (changes requested):
Thanks!

Since we already have logic for this array-merging somewhere else, I think we should refactor to avoid repeating ourselves. There's two reasons for that - both lower code weight, and maintenance benefits if there are bugs in either place.

So I tried this approach that passes the tests:
```diff
diff --git a/ts/Core/Series/DataLabel.ts b/ts/Core/Series/DataLabel.ts
index ffef2466ff..653f40af27 100644
--- a/ts/Core/Series/DataLabel.ts
+++ b/ts/Core/Series/DataLabel.ts
@@ -110,6 +110,10 @@ declare module './SeriesLike' {
             alignTo?: BBoxObject,
             isNew?: boolean
         ): (boolean|undefined);
+        mergeArrays(
+            one: (DataLabelOptions|Array<DataLabelOptions>|undefined),
+            two: (DataLabelOptions|Array<DataLabelOptions>|undefined)
+        ) : (DataLabelOptions|Array<DataLabelOptions>);
         placeDataLabels?(): void;
         setDataLabelStartPos(
             point: ColumnPoint,
@@ -482,6 +486,7 @@ namespace DataLabel {
             seriesProto.alignDataLabel = alignDataLabel;
             seriesProto.drawDataLabels = drawDataLabels;
             seriesProto.justifyDataLabel = justifyDataLabel;
+            seriesProto.mergeArrays = mergeArrays;
             seriesProto.setDataLabelStartPos = setDataLabelStartPos;
             seriesProto.hasDataLabels = hasDataLabels;
         }
diff --git a/ts/Core/Series/Series.ts b/ts/Core/Series/Series.ts
index a63eb55eb2..955c07c6a2 100644
--- a/ts/Core/Series/Series.ts
+++ b/ts/Core/Series/Series.ts
@@ -818,16 +818,10 @@ class Series {
 
         // When plot options have multiple data labels, the default data label
         // plot options are not merged (#21928).
-        const dataLabelOptions = typeOptions.dataLabels;
-        if (dataLabelOptions instanceof Array) {
-            const dataLabelDefaults = defaultPlotOptionsType.dataLabels || {};
-            for (let i = 0; i < dataLabelOptions.length; i++) {
-                dataLabelOptions[i] = merge(
-                    dataLabelDefaults,
-                    dataLabelOptions[i]
-                );
-            }
-        }
+        typeOptions.dataLabels = this.mergeArrays?.(
+            defaultPlotOptionsType.dataLabels,
+            typeOptions.dataLabels
+        );
 
         // Use copy to prevent undetected changes (#9762)
         /**
```

The only thing that doesn't quite sit right here is that `mergeArrays` is a general utility function, but it is attached to the Series prototype for convenience. I think that is okay for now, since it is only used by the data label logic.

Eskils commented:
Thank you for pointing this out. I’ve changed my code to your approach.
